pre-processorのテストエラー修正

@claude 
pre-processorで品質を満たさないコードが書かれています。/pre-processor/app/CLAUDE.mdの実装方針に従いつつ修正して。TDDで。

--------------------

Results:


[/***/***/actions-runner-5/_work-5/Alt/Alt/pre-processor/app/utils/metrics.go:342] - G115 (CWE-190): integer overflow conversion uint64 -> int64 (Confidence: MEDIUM, Severity: HIGH)
    341: 		GCCount:    m.NumGC,
  > 342: 		GCLastTime: time.Unix(0, int64(m.LastGC)),
    343: 	}

Autofix: 

[/***/***/actions-runner-5/_work-5/Alt/Alt/pre-processor/app/utils/optimized_logging.go:179] - G115 (CWE-190): integer overflow conversion int -> uint64 (Confidence: MEDIUM, Severity: HIGH)
    178: 	// Sample based on rate
  > 179: 	if count%uint64(s.samplingRate) == 0 {
    180: 		// Add sampling information

Autofix: 

[/***/***/actions-runner-5/_work-5/Alt/Alt/pre-processor/app/utils/metrics.go:585] - G104 (CWE-703): Errors unhandled (Confidence: HIGH, Severity: LOW)
    584: 	bytes := make([]byte, 8)
  > 585: 	rand.Read(bytes)
    586: 	return fmt.Sprintf("%x", bytes)

Autofix: 

[/***/***/actions-runner-5/_work-5/Alt/Alt/pre-processor/app/utils/metrics.go:579] - G104 (CWE-703): Errors unhandled (Confidence: HIGH, Severity: LOW)
    578: 	bytes := make([]byte, 16)
  > 579: 	rand.Read(bytes)
    580: 	return fmt.Sprintf("%x", bytes)

Autofix: 
make: *** [security] Error 1

Summary:
  Gosec  : dev
  Files  : 63
  Lines  : 10854
  Nosec  : 0
  Issues : 4


---------------------

--- a/pre-processor/app/utils/logger/context_logger.go
+++ b/pre-processor/app/utils/logger/context_logger.go
@@ -10,9 +10,7 @@ import (
 
 type ContextKey string
 
-var (
-	Logger *slog.Logger
-)
+// Logger is imported from global_logger.go
 
 const (
 	RequestIDKey ContextKey = "request_id"

--- a/pre-processor/app/utils/logger/global_logger.go
+++ b/pre-processor/app/utils/logger/global_logger.go
@@ -1,5 +1,12 @@
 package logger
 
+import (
+	"log/slog"
+)
+
+// Logger is the global logger instance
+var Logger *slog.Logger
+
 // init sets up a no-op logger for tests to avoid nil-pointer panics when
 // the application code uses logger.Logger before the main package configures
 // it. Production code still overrides this value in main.go.
@@ -12,7 +19,7 @@ func init() {
 }
 
 // InitGlobalLogger updates the global Logger with UnifiedLogger
-func InitGlobalLogger(config *LoggerConfig) {
+func InitGlobalLogger(config *UnifiedLoggerConfig) {
 	unifiedLogger := NewUnifiedLoggerWithLevel(config.ServiceName, config.Level)
 	Logger = unifiedLogger.logger
 }

--- a/pre-processor/app/utils/metrics.go
+++ b/pre-processor/app/utils/metrics.go
@@ -331,6 +331,18 @@ func (r *RuntimeMetrics) Collect() RuntimeSnapshot {
 	runtime.ReadMemStats(&m)
 	r.lastMemStats = m
 
+	// Safe conversion of uint64 to int64 for GC last time
+	var gcLastTime time.Time
+	if m.LastGC > 0 {
+		// Check for potential overflow when converting uint64 to int64
+		if m.LastGC <= math.MaxInt64 {
+			gcLastTime = time.Unix(0, int64(m.LastGC))
+		} else {
+			// If overflow would occur, use current time as fallback
+			gcLastTime = time.Now()
+		}
+	}
+
 	return RuntimeSnapshot{
 		Timestamp:  time.Now(),
 		Goroutines: runtime.NumGoroutine(),
@@ -339,7 +351,7 @@ func (r *RuntimeMetrics) Collect() RuntimeSnapshot {
 		MemHeapMB:  float64(m.HeapAlloc) / (1024 * 1024),
 		GCPauseMs:  float64(m.PauseNs[(m.NumGC+255)%256]) / 1e6,
 		GCCount:    m.NumGC,
-		GCLastTime: time.Unix(0, int64(m.LastGC)),
+		GCLastTime: gcLastTime,
 	}
 }
 
@@ -576,13 +588,21 @@ func getOrGenerateTraceID(ctx context.Context) string {
 
 func generateTraceID() string {
 	bytes := make([]byte, 16)
-	rand.Read(bytes)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		// Fallback to time-based ID if crypto/rand fails
+		return fmt.Sprintf("%x", time.Now().UnixNano())
+	}
 	return fmt.Sprintf("%x", bytes)
 }
 
 func generateSpanID() string {
 	bytes := make([]byte, 8)
-	rand.Read(bytes)
+	_, err := rand.Read(bytes)
+	if err != nil {
+		// Fallback to time-based ID if crypto/rand fails
+		return fmt.Sprintf("%x", time.Now().UnixNano()&0xFFFFFFFF)
+	}
 	return fmt.Sprintf("%x", bytes)
 }
 

--- a/pre-processor/app/utils/metrics_test.go
+++ b/pre-processor/app/utils/metrics_test.go
@@ -405,6 +405,76 @@ func TestPerformanceReporter(t *testing.T) {
 	}
 }
 
+// TestSecurityIssues tests for security vulnerabilities
+func TestSecurityIssues(t *testing.T) {
+	tests := []struct {
+		name string
+		test func(t *testing.T)
+	}{
+		{
+			name: "should handle uint64 to int64 conversion safely in GC time",
+			test: func(t *testing.T) {
+				collector := NewRuntimeMetrics()
+				
+				// This should not panic or cause overflow
+				snapshot := collector.Collect()
+				
+				// GC last time should be a valid time
+				assert.False(t, snapshot.GCLastTime.IsZero())
+				assert.True(t, snapshot.GCLastTime.Before(time.Now().Add(time.Second)))
+			},
+		},
+		{
+			name: "should handle rand.Read errors properly in generateTraceID",
+			test: func(t *testing.T) {
+				// Test that generateTraceID returns a valid trace ID
+				traceID := generateTraceID()
+				assert.NotEmpty(t, traceID)
+				assert.Len(t, traceID, 32) // 16 bytes * 2 hex chars = 32 chars
+			},
+		},
+		{
+			name: "should handle rand.Read errors properly in generateSpanID",
+			test: func(t *testing.T) {
+				// Test that generateSpanID returns a valid span ID
+				spanID := generateSpanID()
+				assert.NotEmpty(t, spanID)
+				assert.Len(t, spanID, 16) // 8 bytes * 2 hex chars = 16 chars
+			},
+		},
+		{
+			name: "should generate unique trace IDs",
+			test: func(t *testing.T) {
+				ids := make(map[string]bool)
+				
+				// Generate multiple trace IDs and ensure they're unique
+				for i := 0; i < 100; i++ {
+					traceID := generateTraceID()
+					assert.False(t, ids[traceID], "trace ID should be unique")
+					ids[traceID] = true
+				}
+			},
+		},
+		{
+			name: "should generate unique span IDs",
+			test: func(t *testing.T) {
+				ids := make(map[string]bool)
+				
+				// Generate multiple span IDs and ensure they're unique
+				for i := 0; i < 100; i++ {
+					spanID := generateSpanID()
+					assert.False(t, ids[spanID], "span ID should be unique")
+					ids[spanID] = true
+				}
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, tt.test)
+	}
+}
+
 func BenchmarkMetrics(b *testing.B) {
 	b.Run("AppMetrics_Increment", func(b *testing.B) {
 		metrics := NewAppMetrics()

--- a/pre-processor/app/utils/optimized_logging.go
+++ b/pre-processor/app/utils/optimized_logging.go
@@ -344,3 +344,11 @@ func GetOptimizedLogger(component string) *OptimizedLogger {
 	factory := GetGlobalLoggerFactory()
 	return factory.GetLogger(component)
 }
+
+// getEnvOrDefault returns environment variable value or default
+func getEnvOrDefault(key, defaultValue string) string {
+	if value := os.Getenv(key); value != "" {
+		return value
+	}
+	return defaultValue
+}

--- a/pre-processor/app/utils/optimized_logging.go
+++ b/pre-processor/app/utils/optimized_logging.go
@@ -175,8 +175,22 @@ func NewSamplingLogger(logger *OptimizedLogger, samplingRate int) *SamplingLogge
 func (s *SamplingLogger) LogSampled(level string, msg string, args ...interface{}) {
 	count := atomic.AddUint64(&s.counter, 1)
 
-	// Sample based on rate
-	if count%uint64(s.samplingRate) == 0 {
+	// Sample based on rate with safe conversion
+	if s.samplingRate <= 0 {
+		// If sampling rate is invalid, don't log
+		return
+	}
+	
+	// Safe conversion of int to uint64
+	var samplingRate uint64
+	if s.samplingRate > 0 {
+		samplingRate = uint64(s.samplingRate)
+	} else {
+		// If negative, skip sampling
+		return
+	}
+
+	if count%samplingRate == 0 {
 		// Add sampling information
 		sampledArgs := make([]interface{}, 0, len(args)+2)
 		sampledArgs = append(sampledArgs, "sampled_count", count)

--- a/pre-processor/app/utils/optimized_logging_test.go
+++ b/pre-processor/app/utils/optimized_logging_test.go
@@ -465,3 +465,109 @@ func BenchmarkOptimizedLogging(b *testing.B) {
 		})
 	})
 }
+
+// TestSamplingLoggerSecurityIssues tests for security vulnerabilities in sampling logger
+func TestSamplingLoggerSecurityIssues(t *testing.T) {
+	tests := []struct {
+		name string
+		test func(t *testing.T)
+	}{
+		{
+			name: "should handle negative sampling rate safely",
+			test: func(t *testing.T) {
+				config := LogConfig{
+					Level:           "info",
+					Format:          "json",
+					SamplingEnabled: true,
+					SamplingRate:    -1, // negative value
+				}
+				
+				logger := NewOptimizedLogger("test", config)
+				
+				// This should not panic or cause overflow
+				samplingLogger := NewSamplingLogger(logger, -1)
+				assert.NotNil(t, samplingLogger)
+				assert.Equal(t, -1, samplingLogger.samplingRate)
+				
+				// LogSampled should handle negative sampling rate without panic
+				samplingLogger.LogSampled("info", "test message", "key", "value")
+			},
+		},
+		{
+			name: "should handle zero sampling rate safely",
+			test: func(t *testing.T) {
+				config := LogConfig{
+					Level:           "info",
+					Format:          "json",
+					SamplingEnabled: true,
+					SamplingRate:    0,
+				}
+				
+				logger := NewOptimizedLogger("test", config)
+				samplingLogger := NewSamplingLogger(logger, 0)
+				
+				// This should not panic or cause division by zero
+				samplingLogger.LogSampled("info", "test message", "key", "value")
+			},
+		},
+		{
+			name: "should handle large sampling rate safely",
+			test: func(t *testing.T) {
+				config := LogConfig{
+					Level:           "info",
+					Format:          "json",
+					SamplingEnabled: true,
+					SamplingRate:    1000000,
+				}
+				
+				logger := NewOptimizedLogger("test", config)
+				samplingLogger := NewSamplingLogger(logger, 1000000)
+				
+				// This should not cause overflow when converting int to uint64
+				for i := 0; i < 10; i++ {
+					samplingLogger.LogSampled("info", "test message", "key", "value")
+				}
+				
+				assert.Equal(t, uint64(10), samplingLogger.counter)
+			},
+		},
+		{
+			name: "should handle concurrent access with large sampling rate",
+			test: func(t *testing.T) {
+				config := LogConfig{
+					Level:           "info",
+					Format:          "json",
+					SamplingEnabled: true,
+					SamplingRate:    1000000,
+				}
+				
+				logger := NewOptimizedLogger("test", config)
+				samplingLogger := NewSamplingLogger(logger, 1000000)
+				
+				const numGoroutines = 100
+				const logsPerGoroutine = 10
+				
+				var wg sync.WaitGroup
+				wg.Add(numGoroutines)
+				
+				for i := 0; i < numGoroutines; i++ {
+					go func() {
+						defer wg.Done()
+						for j := 0; j < logsPerGoroutine; j++ {
+							samplingLogger.LogSampled("info", "concurrent test", "id", j)
+						}
+					}()
+				}
+				
+				wg.Wait()
+				
+				expected := uint64(numGoroutines * logsPerGoroutine)
+				assert.Equal(t, expected, samplingLogger.counter)
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, tt.test)
+	}
+}

